gh-76785: Multiple Interpreters in the Stdlib (PEP 554) (#18817)
python/cpython: Doc/library/interpreters.rst

:mod:`interpreters` --- High-level Subinterpreters Module
==========================================================

.. module:: interpreters
   :synopsis: High-level SubInterpreters Module.

**Source code:** :source:`Lib/interpreters.py`

> 

Reviewer: Is the module going to be provisional still in 3.9 or is this going to be part of the stable API? If it's going to be provisional, I think it could include a note in the header to explain that.

--------------

> This module provides highlevel tools for working with sub-interpreters,

Reviewer: Based on my reading of the PEP, one of the primary goals for the module is to provide an alternative module of concurrency with a focus on providing a large degree of isolation. Could this be mentioned in the opening paragraph for the module's documentation? Otherwise, the purpose might not be as immediately clear to users.

Also, minor grammar fix:

```suggestion
This module provides high-level tools for working with sub-interpreters,
```

such as creating them, running code in them, or sending data between them.

> It is a wrapper around the low-level :mod:`_interpreters` module.

Reviewer: I don't believe there are any defined documentation sections for an `_interpreters` module, so this wouldn't do anything here AFAIK.

```suggestion
It is a wrapper around the low-level `_interpreters` module.
```

.. versionchanged:: added in 3.9

Interpreter Objects
-------------------

The Interpreter object represents a single interpreter.
.. class:: Interpreter(id)

    The class implementing a subinterpreter object.

    .. method:: is_running()

       Return whether or not the identified interpreter is running.

>        It returns `True` and `False` otherwise.

Reviewer: IMO, this could be better phrased as:

```
Return `True` if the identified interpreter is running.
```

For comparison, see https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.loop.is_running or https://docs.python.org/3/library/concurrent.futures.html#concurrent.futures.Future.cancelled.

    .. method:: destroy()

       Destroy the interpreter. Attempting to destroy the current

>        interpreter results in a `RuntimeError`.

Reviewer: IIRC, it also results in a `RuntimeError` if an interpreter is attempted to be destroyed which is presently running (from my recollection of the C-API function, `interp_destroy()`). Could this also be mentioned?

    .. method:: run(self, src_str, /, *, channels=None):

       Run the given source code in the interpreter. This blocks
       the current thread until done. `channels` should be in
       the form : `(RecvChannel, SendChannel)`.

RecvChannel Objects
-------------------

The RecvChannel object represents a recieving channel.

.. class:: RecvChannel(id)

    This class represents the receiving end of a channel.

    .. method:: recv()

        Get the next object from the channel, and wait if
        none have been sent. Associate the interpreter
        with the channel.

    .. method:: recv_nowait(default=None)

        Like ``recv()``, but return the default result
        instead of waiting.

    .. method:: release()

        Release the channel for the current interpreter.
        By default both ends are released. Releasing an already

>         released end results in a ``ChannelReleasedError`` exception.

Reviewer: IMO, the second sentence could be better phrased as:

```
Releasing an end more than once results in a
:exc:`ChannelReleasedError`.
```
(Also added a Sphinx role, which should generate an inline link to the exception definition below)

    .. method:: close(force=False)

        Close the channel in all interpreters. By default

>         both ends are closed. closing an already closed end

Reviewer: ```suggestion
        both ends are closed. Closing an already closed end
```

>         results in a ``ChannelClosedError`` exception. Without

Reviewer: ```suggestion
        results in a :exc:`ChannelClosedError` exception. Without
```

>         seeting ``force`` to ``True`` a ``ChannelNotEmptyError``

Reviewer: Grammar fix and changed parameter name to use italics instead of backticks. Also adds sphinx role.

```suggestion
        setting *force* to ``True``, a :exc:`ChannelNotEmptyError`
```

>         will be returned when a channel with data is closed.

Reviewer: ```suggestion
        will be raised when a channel with data is closed.
```

SendChannel Objects
--------------------

The ``SendChannel`` object represents a sending channel.

.. class:: SendChannel(id)

    This class represents the sending end of a channel.

    .. method:: send(obj)

       Send the object ``obj`` to the receiving end of the channel
       and wait. Associate the interpreter with the channel.

    .. method:: send_nowait(obj)

>         Like ``send()`` but return ``False`` if not received.

Reviewer: Phrasing suggestion and clarification:

```suggestion
        Similar to ``send()``, but returns ``False`` if
        *obj* is not immediately received instead of blocking.
```

    .. method:: send_buffer(obj)

       Send the object's buffer to the receiving end of the
       channel and wait. Associate the interpreter with the
       channel.

    .. method:: send_buffer_nowait(obj)

>        Like ``send_buffer()`` but return ``False`` if not received.

Reviewer: See my above comment on `send_nowait()`.

    .. method:: release()

        Release the channel for the current interpreter.
        By default both ends are released. Releasing an already

>         released end results in a ``ChannelReleasedError`` exception.

Reviewer: ```suggestion
        released end results in a :exc:`ChannelReleasedError`.
```

    .. method:: close(force=False)

        Close the channel in all interpreters. By default

>         both ends are closed. closing an already closed end

Reviewer: ```suggestion
        both ends are closed. Closing an already closed end
```

>         results in a ``ChannelClosedError`` exception.

Reviewer: ```suggestion
        results in a :exc:`ChannelClosedError`.
```

This module defines the following global functions:

.. function:: is_shareable(obj)

   Return ``True`` if the object's data can be shared between
   interpreters.

.. function:: create_channel()

   Create a new channel for passing data between interpreters.

.. function:: list_all_channels()

>    Return all open channels.

Reviewer: Does this return a list of Channel objects or something else? I think it would be helpful to mention that here.

.. function:: create()

   Initialize a new (idle) Python interpreter. Get the currently

>    running interpreter. This method returns an ``Interpreter`` object.

Reviewer: Was "Get the currently running interpreter" accidentally included here?
